packages/langium/src/documents/document-builder.ts

```diff
@@ -54,55 +50,72 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
         this.indexManager = services.index.IndexManager;
     }
 
-    build(document: LangiumDocument): BuildResult {
-        const parseResult = this.parser.parse(document);
-        document.parseResult = parseResult;
-        this.process(document);
-        let diagnostics: Diagnostic[] | undefined = this.connection ? this.validate(document) : undefined;
-        const validator = this.documentValidator;
+    async build(document: LangiumDocument, cancelToken = CancellationToken.None): Promise<BuildResult> {
+        await this.buildDocument(document, cancelToken);
         return {
             parseResult: document.parseResult,
-            get diagnostics() {
-                if (!diagnostics) {
-                    diagnostics = validator.validateDocument(document);
-                }
-                return diagnostics;
-            }
+            diagnostics: await this.validate(document, cancelToken, true)
         };
     }
 
-    validate(document: LangiumDocument): Diagnostic[] | undefined {
-        let diagnostics: Diagnostic[] | undefined;
+    protected async validate(document: LangiumDocument, cancelToken = CancellationToken.None, forceDiagnostics = false): Promise<Diagnostic[]> {
+        let diagnostics: Diagnostic[] = [];
         const validator = this.documentValidator;
-        if (this.connection) {
-            diagnostics = validator.validateDocument(document);
-            // Send the computed diagnostics to VS Code.
-            this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+        if (this.connection || forceDiagnostics) {
+            diagnostics = await validator.validateDocument(document, cancelToken);
+            if (this.connection && !cancelToken.isCancellationRequested) {
+                // Send the computed diagnostics to VS Code.
+                this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+            }
+        }
+        if (!cancelToken.isCancellationRequested) {
+            document.state = DocumentState.Validated;
         }
         return diagnostics;
     }
 
-    documentChanged(uri: URI): void {
+    async documentChanged(uri: URI, cancelToken = CancellationToken.None): Promise<void> {
         this.langiumDocuments.invalidateDocument(uri);
         const newDocument = this.langiumDocuments.getOrCreateDocument(uri);
-        this.indexManager.update(newDocument);
-        this.langiumDocuments.all.filter(doc => this.isAffected(doc, uri)).forEach(
-            doc => {
-                this.build(doc);
-                this.indexManager.update(doc);
+        await delayNextTickAfter();
+        if (cancelToken.isCancellationRequested) {
```

Comment:
Thought: what if we provide a special cancellation token where `isCancellationRequested` is of type `Promise<boolean>` and that internally calls `delayNextTickAfter` through a getter?
@msujew WDYT?

Reply:
Sure, sounds good to me!

Reply:
The advantage is that the caller of all these API methods can control themselves whether to interrupt the execution and after which time period.

The drawback is that the token interface is not compatible with that of `vscode-languageserver`. We could translate `vscode-languageserver` CancellationTokens into `langium` CancellationTokens with a utility function. Should we use a different name to avoid clashes?

Reply:
For now, I thought about a `CancelToken` class that simply accepts a `CancellationToken` in its constructor. Or should we separate this stronger using an interface?

Reply:
I would prefer an interface, and it could be an extension of CancellationToken that adds a property. Example:
```typescript
export interface AsyncCancellationToken extends CancellationToken {
    readonly interruptAndCheck: Promise<boolean>;
}
```

Another thought, what do you think of this:
```typescript
export const OperationCanceled = Symbol('OperationCanceled');

export function interruptAndCheck(token: AsyncCancellationToken): Promise<void> {
    if (await token.interruptAndCheck) {
        throw OperationCanceled;
    }
}
```

This is similar to how it's done in Xtext:
https://github.com/eclipse/xtext-core/blob/83ae598a64313f511a2b2993dc23f0d17a092c9c/org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java#L85-L89

It's a way to avoid the frequent `if (...) { return; }` blocks.

Reply:
Sure, sounds good as well 👍

Reply:
It's probably not a good idea to choose the same name for the property and the utility function, though...